Issue/2289 read only shipping labels

anitaa1990 opened the pull request:
Fixes #2289. This PR adds the UI to display read only shipping labels of an order in the Order detail screen.

Apologies in advance for the detailed testing in this PR 🙇‍♀️
~**This PR is in draft till this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1609) can be reviewed and merged.**~

#### Notes
There have been a lot of UI changes to the Order detail screen because of this change and I would like to describe the most important ones below.
- The shipping labels for an order includes products associated with the shipping label, the tracking number of the shipment and the package details. This means: 
  - We no longer display shipment tracking list of an order separately in Order detail screen.
  - We no longer provide the option to add a tracking list in the app if shipping labels are available. cc @Garance91540 I remember we discussed that so I had implemented it that way. Please let me know if we need to change this. 
  - We don't display the product list of an order separately. The products associated with a shipping label is displayed as part of the shipping label. Only products that are no associated with any shipping labels are displayed separately in the product list.
- If a shipping label has been refunded, we no longer provide the option to track the shipment link and mark the shipping label as refunded (see below screenshot).

#### Screenshots
##### Shipping label list
<img src="https://user-images.githubusercontent.com/22608780/85259296-6fa62c00-b486-11ea-918f-e7c959306df8.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/85259301-716fef80-b486-11ea-8416-bdbb8c0f0d5d.png" width="300"/>

##### Multi package shipping label list
<img src="https://user-images.githubusercontent.com/22608780/85259891-708b8d80-b487-11ea-8f7a-c9e93329a1cd.gif" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/85259899-72555100-b487-11ea-9bbc-7c49ba30ebac.gif" width="300"/>

##### Shipping label list details
<img src="https://user-images.githubusercontent.com/22608780/85260009-9d3fa500-b487-11ea-9ad9-91657a51fe1e.gif" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/85260020-9fa1ff00-b487-11ea-89e9-219f0f3dc3ce.gif" width="300"/>

##### Refunded shipping labels
<img src="https://user-images.githubusercontent.com/22608780/85260063-b5afbf80-b487-11ea-9040-bfa7b8fa3752.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/85260073-b8aab000-b487-11ea-9dc6-7ff3389b0dae.png" width="300"/>

#### Testing
Follow this guide on `p91TBi-2xo` to add a test credit card to your WordPress.com login. 

**Scenario I: 1 Shipping label for an order with only one product and with the order fulfilled automatically.** 
- Create an order from your test site with only one physical product.
- Create a shipping label for that order from your test site. Ensure that you check the box that says fulfil the order automatically.
- Open the app and click on that order.
    - Verify that the shipping label is displayed correctly (header should be Products or Package 1?!)
    - Under the header, the product (along with it’s quantity, price) should be displayed.
    - Below that, the tracking number with the service name should be displayed.
    - Clicking on the tracking icon should open the shipment tracking website.
    - Click on `Show Shipment Details` and verify that the shipping details are displayed correctly. 
    - Notice that the `Begin Fulfillment` button is NOT displayed.
- Request a refund for the shipping label for this order. Open the app and click on that order.
    - Verify that the tracking icon is no longer displayed.
    - The tracking number is displayed along with the refund status and the time of refund.
    - If the order is NOT fulfilled, display the begin fulfilment button. 

**Scenario II: 1 Shipping label for an order with only one product and with the order fulfilled manually.** 
- Create an order from your test site with only one physical product.
- Create a shipping label for that order from your test site. Ensure that you uncheck the box that says fulfil the order automatically.
- Open the app and click on that order.
    - Verify that the shipping label is displayed correctly (header should be Products or Package 1?!)
    - Under the header, the product (along with it’s quantity, price) should be displayed.
    - Below that, the tracking number with the service name should be displayed.
    - Clicking on the tracking icon should open the shipment tracking website.
    - Click on `Show Shipment Details` and verify that the shipping details are displayed correctly. 
    - Notice that the `Begin Fulfillment` button is displayed and clicking on it redirects to the Order Fulfill screen.
- Request a refund for the shipping label for this order. Open the app and click on that order.
    - Verify that the tracking icon is no longer displayed.
    - The tracking number is displayed along with the refund status and the time of refund.
    - If the order is NOT fulfilled, display the begin fulfilment button.

**Scenario III: 2 Shipping labels for an order with 2 products and with the order fulfilled automatically.**  
- Create an order from your test site with only 2 physical products.
- Create 2 separate shipping labels for each of the products, from your test site. Ensure that you check the box that says fulfil the order automatically.
- Open the app and click on the order.
    - Verify that 2 shipping labels are displayed correctly. Header should be (Package 1 for the first shipping label and Package 2 for the second shipping label).
    - Under the header, the product associated with the shipping label should be displayed correctly.
    - Below that, the tracking number with the service name should be displayed.
    - Clicking on the tracking icon should open the shipment tracking website.
    - Click on `Show Shipment Details` and verify that the shipping details are displayed correctly. 
    - Notice that the `Begin Fulfillment` button is NOT displayed.
- Request a refund for one of the shipping labels for this order. Open the app and click on that order.
    - Verify that for that shipping label: 
        - the tracking icon is no longer displayed.
        - The tracking number is displayed along with the refund status and the time of refund.
    - Verify that for the shipping label that is NOT refunded, the details are displayed correctly. 

**Scenario IV: 2 Shipping labels for an order with 2 products with the order fulfilled manually.**  
- Create an order from your test site with two physical products.
- Create 2 separate shipping labels for each of the products, from your test site. Ensure that you uncheck the box that says fulfil the order automatically.
- Open the app and click on the order.
    - Verify that 2 shipping labels are displayed correctly. Header should be (Package 1 for the first shipping label and Package 2 for the second shipping label).
    - Under the header, the product associated with the shipping label should be displayed correctly.
    - Below that, the tracking number with the service name should be displayed.
    - Clicking on the tracking icon should open the shipment tracking website.
    - Click on `Show Shipment Details` and verify that the shipping details are displayed correctly. 
    - Notice that the `Begin Fulfillment` button is displayed. But not sure where we should display this button? Under Package 1 or Package 2?!
- Request a refund for one of the shipping labels for this order. Open the app and click on that order.
    - Verify that for that shipping label: 
        - the tracking icon is no longer displayed.
        - The tracking number is displayed along with the refund status and the time of refund.
    - Verify that for the shipping label that is NOT refunded, the details are displayed correctly.

**Scenario V: 2 Shipping labels for an order with more than 2 products with the order fulfilled manually.**  
- Create an order from your test site with at least 3 products.
- Create 2 separate shipping labels for two of the products, from your test site. Ensure that you remove one of thuncheck the box that says fulfil the order automatically.
- Ensure that one of the products from the order is not associated with any of the shipping labels.
- Open the app and click on the order.
    - Verify that 2 shipping labels are displayed correctly. Header should be (Package 1 for the first shipping label and Package 2 for the second shipping label).
    - Verify that a third section is displayed with Header as “Remaining Items to ship” and it displays the order that is not associated with any of the shipping labels. 
    - Notice that the `Begin Fulfillment` button is displayed. 

**Scenario VI: Order detail for an order without shipping labels.**  
- Open the app and click on an order that does not include any shipping labels.
- Verify that the order detail is displayed as expected.

**Scenario VII: Order detail when shipping label i.e. WCS is not available.**  
- Login to a store with WCS not available. Click on any order from the app.
- Verify that the order detail is displayed as expected.

**Scenario VIII: Order detail for a virtual product.** 
- Open the app and click on an order for a virtual product. We should not be displaying shipping labels for virtual products.
- Verify that the order detail is displayed as expected.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Garance91540 commented:
> Follow this guide on p91TBi-2xo to add a test credit card to your WordPress.com login.

Could you tell me more on how to do this? Can I test all the changes in this PR by downloading the APK?

Garance91540 commented:
@anitaa1990  I downloaded the APK and started playing with the orders, but couldn't see my shipping labels or my refunded labels. Is this the right apk file?

anitaa1990 commented:
@Garance91540 ,
> Follow this guide on p91TBi-2xo to add a test credit card to your WordPress.com login.
Could you tell me more on how to do this? Can I test all the changes in this PR by downloading the APK?

I sent you the link to a fieldguide via DM that will let you add a dummy credit card to your account so that can purchase shipping labels for an order. 

> @anitaa1990 I downloaded the APK and started playing with the orders, but couldn't see my shipping labels or my refunded labels. Is this the right apk file?

Hmm, it should be 🤔 In any case, I shared the apk file directly via DM. 

Let me know if you face any issues!

Thank you for confirming the scenarios from my notes 😄

Garance91540 commented:
HI @anitaa1990! With the APK file you sent me in DM it works, thank you! There are a lot of things to review here so I'll probably do several rounds! (Maybe in the future we can break down the PRs to tackle these cases separately? I know it's probably easier said than done :D) 

**Manually fulfilled orders**

The desired behavior for the Product card header is by default to display "Products", and once the order is fulfilled, if there are still products which haven't been added to a package, then we'll display "Remaining items to ship" along with "Package 1"/"Package 2".

**Refunded label**

I added "V1" designs in Figma and tagged you directly, it might be easier to review there. 

<img width="660" alt="Screen Shot 2020-06-23 at 12 12 05 PM" src="https://user-images.githubusercontent.com/28450084/85391684-caa15700-b54a-11ea-883e-9cd93c90799d.png">

**Read-only label**

Screenshot:
![Screenshot_20200623-121923](https://user-images.githubusercontent.com/28450084/85392514-ece7a480-b54b-11ea-9b39-0735ff853eb4.png)

Tracking number: 
is it possible to add the carrier before the tracking number? So here it should read "USPS Tracking number" instead of "Tracking number". I'm also not seeing the "External link" icon that redirects to the tracking website?

Shipment details:
Is it possible to update the vertical padding of the icon of each section (ship from, ship to, package details, etc...) to 16px?

<img width="683" alt="Screen Shot 2020-06-23 at 12 22 31 PM" src="https://user-images.githubusercontent.com/28450084/85392816-59fb3a00-b54c-11ea-8bf6-b8ccd5047dd3.png">

anitaa1990 commented:
Thanks for the review @Garance91540!!

> The desired behavior for the Product card header is by default to display "Products", and once the order is fulfilled, if there are still products which haven't been added to a package, then we'll display "Remaining items to ship" along with "Package 1"/"Package 2".

That makes sense. I started with that but there were scenarios where the order has multiple shipping label packages for each of the products and the order is not fulfilled. In this case we would still display it as `Package 1` and `Package 2` right? That’s why I had added Package 1, package 2 etc.

But if the order has only a single package, we can keep it as `Products`. WDYT? 

Also, the option to fulfil an order is kept as an optional field in the web. 
<img width="1282" alt="Screenshot 2020-06-24 at 6 01 02 PM" src="https://user-images.githubusercontent.com/22608780/85556901-e9841400-b644-11ea-8274-410f3ee7d983.png">


So if there are multiple shipping labels for an order and if the user chose not to fulfill the order, where should we display the Fulfill Order button in the app? Under Package 1 or Package 2?

> is it possible to add the carrier before the tracking number? So here it should read "USPS Tracking number" instead of "Tracking number". I'm also not seeing the "External link" icon that redirects to the tracking website?

Yup, for sure! I'll work on that and send you another APK file for testing!

I couldn't add comments to the Figma file so ended up asking questions here 😅

Garance91540 commented:
Hi @anitaa1990 ! Here's my review of each of the scenario mentioned!

**Scenario 0: Order Details screen - Order not fulfilled yet**

Products card header should read "Products" (currently reading "Remaining items to ship)

**Scenario I: 1 Shipping label for an order with only one product and with the order fulfilled automatically.**

To answer your question here:

> Verify that the shipping label is displayed correctly (header should be Products or Package 1?!)

Header should be "Package 1" 

test: ✅ 

(Nothing to say except the UI changes I mentioned in Figma (update icons alignment in the expanded shipment details view, add carrier name in "Tracking number", show bottom divider of the last product in the list, add "label" when label is refunded.)

label refund: could not test, I have an error message when I try to refund my label

**Scenario II: 1 Shipping label for an order with only one product and with the order fulfilled manually.**

test: ✅  (Products header correctly displays "Package 1")

label refund: could not test, I have an error message when I try to refund my label

**Scenario III: 2 Shipping labels for an order with 2 products and with the order fulfilled automatically.**
test: ✅ 
label refund: could not test, I have an error message when I try to refund my label

anitaa1990 commented:
Thanks for the review @Garance91540!!

> 1. Products card header should read "Products" (currently reading "Remaining items to ship).
2. add carrier name in "Tracking number", 
3. show bottom divider of the last product in the list, 
4. add "label" when label is refunded. 

All the above feedback has been addressed! Please see below screenshots and let me know your feedback. Thank you 😄 
##### Single Package Shipping label
<img src="https://user-images.githubusercontent.com/22608780/85952883-f4fb7600-b989-11ea-84d0-4f03247f7642.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/85952884-f75dd000-b989-11ea-8127-962f02ed9dc8.png" width="300"/>. 

##### Multi Package Shipping label
<img src="https://user-images.githubusercontent.com/22608780/85952887-faf15700-b989-11ea-91a3-ffae83179f5b.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/85952889-fd53b100-b989-11ea-8083-62a2a2806ce5.png" width="300"/>. 

#### Refunded Shipping label
<img src="https://user-images.githubusercontent.com/22608780/85952893-ffb60b00-b989-11ea-8611-244bae34c1a4.png" width="300"/>  

> update icons alignment in the expanded shipment details view

The current icon is aligned at 16px only 🤔 16px top, bottom and left of the icons.

I have also sent you the apk file via DM for you to test 👍  Let me know if you face any issues!

anitaa1990 commented:
Thanks for the review @Garance91540!

> Ah I think there was some miscommunication there :D ! Here are the designs for this row: https://d.pr/i/swcM5W. This section should be 2-line:
first line: header= "carrier name" + "tracking number" (ie: "USPS Tracking number") and
second line= tracking number

Ah, my bad! I've modified it in the below screenshot!
<img src="https://user-images.githubusercontent.com/22608780/86006928-5841e280-ba34-11ea-82b4-c7bd268fcf91.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/86006935-5a0ba600-ba34-11ea-9eb7-c7a0d6ac2ee1.png" width="300"/>

> The current icon is aligned at 16px only 🤔 16px top, bottom and left of the icons. Weird! here is a screenshot of what I see compared to the design. As you can see, the icons on the screenshots are not top aligned. 

I had set the icon to be center aligned within the view when it should be top aligned! Thanks! Fixed in the above screenshot!

anitaa1990 commented:
Thanks for the review @Garance91540!!

>  for the tracking number, can we just display "Carrier name" + "tracking number"? (No need to display the name of the label again since it's in the shipment details)

Unfortunately, we don't get the Carrier name separately from the API. We get the package name which includes the carrier name as well and we get something called the `carrier_id` which just displays "usps". But in both cases, the API side could change that at any point and it might not be good to add extra logic in the app to parse the carrier name from the package name. I'll try to find out if there is anyway to just get the carrier name here after some more extra debugging!

Garance91540 commented:
Ah gotcha, I didn't know that was the case! If we are not able to just get the carrier name, then let's just display "Tracking number" without anything else!

AmandaRiu commented:
@anitaa1990 Also, all the tests in `OrderFulfillmentProductListCardTest` are failing for me with similar errors to this:

```
androidx.test.espresso.PerformException: Error performing 'single click - At Coordinates: 539, 611 and precision: 16, 16' on view 'with id: com.woocommerce.android.dev:id/ordersList'.
```

Are you seeing something similar on your end?

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/StringExt.kt:
```diff
@@ -58,3 +59,14 @@ fun String.fastStripHtml(): String {
+fun String.getCountryLabelByCountryCode(): String {
+    val locale = Locale(Locale.getDefault().language, this)
+    return locale.displayCountry
+}
```
Good point! I removed the extension method and updated the code to use the method in `AddressUtils` in 7791d7b

anitaa1990 commented:
Thanks for the thorough review @AmandaRiu!

>  There is an order (9120 for my testwooshop store) for only a single product. I created a shipping label for it in wp-admin so there is just one package and the order is now "complete". When I try to open the order in the app the app will immediately crash and here is the error from the log:

Nice catch! This fix requires an update in FluxC so I opened a [PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1635) for that and updated the hash in b659273.

> Also, all the tests in OrderFulfillmentProductListCardTest are failing for me

I believe most of the UI tests are outdated and failing at the moment, even in `develop`. After the refactor to the Product detail, Order list and Order detail, some of those tests are redundant and there was some discussion to remove them altogether. I think the only tests that are working are the ones used for Screenshots and that's part of CircleCI (which is passing) so I think it's safe to ignore those tests for now. 

Ready for another round 👍

AmandaRiu commented:
That was weird. Approving and merging the FluxC PR actually closed this PR 😩 Reopened.

Garance91540 commented:
Hi @AmandaRiu , here are some answers to your comments:

> What I noticed is the virtual item is now in a card titled "Remaining items to ship" but that doesn't make much sense since you wouldn't ship a virtual item. 

Good point, I hadn't thought of that scenario! @anitaa1990 is it hard to change "Remaining items to ship" with "Remaining products" if they are virtual?

> Also the details button makes it seem like maybe you would view details of these unshipped items but instead you view all items:

I actually created a separate PR in the past to request we delete the "Details" section. I'm not sure it's adding much to the experience, you can see all the details either on the Order page or when selecting the products. What would you think of removing it? I'm also not seeing it anymore on iOS.

anitaa1990 commented:
> What I noticed is the virtual item is now in a card titled "Remaining items to ship" but that doesn't make much sense since you wouldn't ship a virtual item.
Good point, I hadn't thought of that scenario! @anitaa1990 is it hard to change "Remaining items to ship" with "Remaining products" if they are virtual?

This is definitely doable @Garance91540  for scenarios where we display only virtual products with the header as "Remaining products". But what happens in those scenarios where there are some physical products not associated with any shipping label and virtual products as well? Are we planning to display both of those in separate cards? That is also doable but needs some extra work to happen in the app. I would prefer tackling that in a separate PR, if that ok with both of you? For now, I just updated the logic in the app to display "Remaining products" as the header if all the products not associated with any shipping label, are virtual products.

> if WCS was not installed, the products card would sometimes just not display in the order details. There didn't seem to be a lot of rhyme or reason as to why some would show up just fine while others would not, but it did seem like orders with more products were more likely to not show the products card (but not always). I'm wondering if it's some sort of race condition in the code.

Thanks for the thorough review @AmandaRiu!!  That was an excellent catch! And the screenshot of the order helped me to recreate the issue. It was because we are only displaying the products section if there are no shipping labels and if none of the products are refunded. I had modified this logic to check for refunded products and it was causing the issue. I fixed that in f09bcf2 and it should work correctly now 🤞 Let me know if you face any issues!!

>  Also, if all the products in an order have been added to a package there is not "Details" button to view a full list of the ordered items.

As G mentioned, I think we decided to not display the "Details" button since all the information displayed in the details screen were actually being displayed in the order detail screen so it was deemed redundant.

Ready for another round 👍

anitaa1990 commented:
> What if I were to have an order with refunded products AND multiple packages. I tried it out and ran into an issue.

Eek, yeah I was afraid of something like this! So the API response for the shipping labels send us only the product names associated with each shipping label. In the app, we get the product details for that product name to display the product image, quantity, price etc. Since both the `Variable Hoodie` have the same name, it is displayed as part of both packages. I think we might need the productId or the variation id of the product from the API side in order to tackle this scenario. In the web, only the product names as such is displayed so this isn't a problem. I'll talk with the API team and come up with a workaround for this in a subsequent PR. Since this is behind a feature flag it shouldn't affect users. WDYT?

Thanks for the thorough review 🙇‍♀️ Ready for another round 👍

astralbodies commented:
Moving this to the 4.8 release.

Garance91540 commented:
> This is definitely doable @Garance91540 for scenarios where we display only virtual products with the header as "Remaining products". But what happens in those scenarios where there are some physical products not associated with any shipping label and virtual products as well? Are we planning to display both of those in separate cards? That is also doable but needs some extra work to happen in the app. I would prefer tackling that in a separate PR, if that ok with both of you? For now, I just updated the logic in the app to display "Remaining products" as the header if all the products not associated with any shipping label, are virtual products.

🤯🤯🤯 I think for V1 "Remaining products" works great! Let's forget "Remaining products to ship" for now completely

oguzkocer commented:
@astralbodies Since this is merged before the code freeze, should the milestone be `4.7`? I don't want to update the milestone directly just in case I am missing something.

AmandaRiu commented:
@oguzkocer This is fine for milestone 4.6, @astralbodies only changed it because he didn't think it'd make it before the cut. I'll change it back.
